### PR TITLE
make mock driver support PV healthy monitor

### PR DESCRIPTION
**Pull request by fengzixu**
#### What I did in this PR

1. Mock ControllerGetVolume RPC interface
2. Extend ListVolumes for supporting VolumeCondition support
3. Extend the NodeVolumeStats functions

This PR is for supporting test of PV health monitor
Fixes: https://github.com/kubernetes-csi/external-health-monitor/issues/9

```release-note
NONE
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,7 +3,7 @@ module github.com/kubernetes-csi/csi-test/v3
 go 1.12
 
 require (
-	github.com/container-storage-interface/spec v1.2.0
+	github.com/container-storage-interface/spec v1.2.0-rc1.0.20200530102742-4c5274ab4e5e
```
Comment (xing-yang) [inline go.mod:6]: You can use this one:
https://github.com/container-storage-interface/spec/releases/tag/v1.3.0-rc1

File: mock/service/controller.go [hunk 2]
```diff
@@ -326,6 +326,38 @@ func (s *service) ValidateVolumeCapabilities(
...
+	if i < 0 {
+		return nil, status.Error(codes.NotFound, req.VolumeId)
+	}
+
+	resp := &csi.ControllerGetVolumeResponse{
+		Volume: &v,
+		Status: &csi.ControllerGetVolumeResponse_VolumeStatus{},
+	}
+
+	volumeStatus, err := s.NodeGetVolumeStats(ctx, &csi.NodeGetVolumeStatsRequest{})
```
Comment (xing-yang) [inline mock/service/controller.go:348]: Why is NodeGetVolumeStats called within ControllerGetVolume?

File: mock/service/controller.go [hunk 3]
```diff
@@ -388,8 +420,19 @@ func (s *service) ListVolumes(
 	)
 
 	for i = 0; i < len(entries); i++ {
+		volumeStatus, err := s.NodeGetVolumeStats(ctx, &csi.NodeGetVolumeStatsRequest{
```
Comment (xing-yang) [inline mock/service/controller.go:423]: ListVolumes is controller PRC.  We should not call a Node PRC within a controller RPC.

**Comment by NickrenREN** [comment]
>What I did in this PR
Mock ControllerGetVolume RPC interface
Extend ListVolumes for supporting VolumeCondition support

There are two newly added RPC interfaces, one of controller side: `ControllerGetVolume` and the other one is node side: `NodeGetVolumeStats`. Both of them should be mocked here

File: go.mod (hunk 1 above)
Comment (fengzixu) [reply go.mod]: Got it.

File: mock/service/controller.go (hunk 2 above)
Comment (fengzixu) [reply mock/service/controller.go]: I misunderstand the working mechanism between Controller and Agent. Revised

File: mock/service/controller.go (hunk 3 above)
Comment (fengzixu) [reply mock/service/controller.go]: Sure. Thanks for your comment.

**Comment by fengzixu** [comment]
@xing-yang @NickrenREN 

Please review it again.

File: driver/driver.mock.go [hunk 4]
```diff
@@ -234,6 +234,13 @@ func (m *MockControllerServer) ListVolumes(arg0 context.Context, arg1 *csi.ListV
 	return ret0, ret1
 }
 
+func (m *MockControllerServer) ControllerGetVolume(arg0 context.Context, arg1 *csi.ControllerGetVolumeRequest) (*csi.ControllerGetVolumeResponse, error) {
+	ret := m.ctrl.Call(m, "ControllerGetVolume", arg0, arg1)
+	ret0, _ := ret[0].(*csi.ControllerGetVolumeResponse)
+	ret1, _ := ret[1].(error)
+	return ret0, ret1
+}
+
```
Comment (NickrenREN) [inline driver/driver.mock.go:243]: Can you mock Node method too ?

Comment (NickrenREN) [inline driver/driver.mock.go:243]: Could you mock Node method too ?

Comment (NickrenREN) [reply driver/driver.mock.go:243]: Oh, it seems `NodeGetVolumeStats` is already there

File: mock/service/controller.go [hunk 5]
```diff
@@ -326,6 +326,32 @@ func (s *service) ValidateVolumeCapabilities(
...
+	req *csi.ControllerGetVolumeRequest) (
+	*csi.ControllerGetVolumeResponse, error) {
+
+	if hookVal, hookMsg := s.execHook("GetVolumeStart"); hookVal != codes.OK {
+		return nil, status.Errorf(hookVal, hookMsg)
+	}
+
+	i, v := s.findVolByID(ctx, req.VolumeId)
+	if i < 0 {
+		return nil, status.Error(codes.NotFound, req.VolumeId)
```
Comment (NickrenREN) [inline mock/service/controller.go:340]: Could we return `VolumeCondition` and set its `Abnormal` and `Message` fields

File: mock/service/controller.go [hunk 6]
```diff
@@ -326,6 +326,32 @@ func (s *service) ValidateVolumeCapabilities(
...
+	}
+
+	resp := &csi.ControllerGetVolumeResponse{
+		Volume: &v,
+		Status: &csi.ControllerGetVolumeResponse_VolumeStatus{},
```
Comment (NickrenREN) [inline mock/service/controller.go:345]: Could we generate the `VolumeCondition` field instead of returning an empty one (nil)?

File: mock/service/node.go [hunk 7]
```diff
@@ -427,6 +434,7 @@ func (s *service) NodeGetVolumeStats(ctx context.Context,
 				Unit:  csi.VolumeUsage_BYTES,
 			},
 		},
+		VolumeCondition: &csi.VolumeCondition{},
```
Comment (NickrenREN) [inline mock/service/node.go:437]: Could we also construct some abnormal conditions ?

File: mock/service/controller.go (hunk 6 above)
Comment (NickrenREN) [reply mock/service/controller.go:345]: Please refer to the comments
```
type NodeGetVolumeStatsResponse struct {
	// This field is OPTIONAL.
	Usage []*VolumeUsage `protobuf:"bytes,1,rep,name=usage,proto3" json:"usage,omitempty"`
	// Information about the current condition of the volume.
	// This field is OPTIONAL.
	// This field MUST be specified if the VOLUME_CONDITION node
	// capability is supported.
	VolumeCondition      *VolumeCondition
...
```

Comment (fengzixu) [reply mock/service/controller.go]: For error, I mock the VolumeCondition filed. For normal response, I mocked the PublishedNodeIds

**Comment by fengzixu** [comment]
@xing-yang @NickrenREN Request review

File: mock/service/controller.go (hunk 6 above)
Comment (NickrenREN) [reply mock/service/controller.go]:
```       
        // Information about the current condition of the volume.
	// This field is OPTIONAL.
	// This field MUST be specified if the
	// VOLUME_CONDITION controller capability is supported.
```

File: mock/service/controller.go [hunk 8]
```diff
@@ -326,6 +326,41 @@ func (s *service) ValidateVolumeCapabilities(
...
+		Status: &csi.ControllerGetVolumeResponse_VolumeStatus{
+			VolumeCondition: &csi.VolumeCondition{},
+		},
+	}
+	i, v := s.findVolByID(ctx, req.VolumeId)
+	if i < 0 {
+		resp.Status.VolumeCondition.Abnormal = true
+		resp.Status.VolumeCondition.Message = "volume not found"
+		return resp, status.Error(codes.NotFound, req.VolumeId)
+	}
```
Comment (xing-yang) [inline mock/service/controller.go:348]: Can you add a case where volume is found but condition is Abnormal=true?

Comment (fengzixu) [reply mock/service/controller.go:348]: For the mock test, I think we can divide it into 2 parts to do
1. Add basic mock test functions for testing if volume-healthy-monitor can work(Due to the initial implementation PR wasn’t merged, it is hard to do further works)
2. Add more unit, integration, e2e test for testing volume-healthy-monitor can work correctly on positive and negative cases

The reasons I want to do a testing task like above are below
1. The initial PR blocks many works and it isn’t perfect. We can ensure it works normally and improves it
2. After initial PR was merged. It is better for us to let more contributors contribute this project.

Comment (xing-yang) [reply mock/service/controller.go:348]: Sure
